### PR TITLE
libelfin: reading ELF binaries and DWARFv4 debug information

### PR DESCRIPTION
**Pull request by klimkin**
C++11 library for reading ELF binaries and DWARFv4 debug information.

Specify library name and version:  **libelfin/0.3**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libelfin/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+sources:
+  "0.2":
+    url: "https://github.com/aclements/libelfin/archive/v0.2.tar.gz"
+    sha256: "ecbe3d482697245ac6fbac7dfd24e4e5b37b76cd860592e94385620e3238a179"
```
Comment (madebr) [inline recipes/libelfin/all/conandata.yml:4]: Unless really needed, don't add older releases. (also remove from `config.yml`)

File: recipes/libelfin/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if self.options.shared:
+            del self.options.fPIC
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def build(self):
+        for patch in self.conan_data["patches"][self.version]:
```
Comment (madebr) [inline recipes/libelfin/all/conanfile.py:36]: Version `0.2` does not contain a `patches` section in `conandata.yml` and will fail.
```suggestion
        for patch in self.conan_data.get("patches", {}).get(self.version, []):
```

File: recipes/libelfin/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            tools.patch(**patch)
+        with tools.chdir(self._source_subfolder):
+            autotools = AutoToolsBuildEnvironment(self)
+            autotools.make()
+            autotools.install(args=["PREFIX=" + os.path.join(self.source_folder, "package")])
```
Comment (madebr) [inline recipes/libelfin/all/conanfile.py:41]: Please check other recipes using `AUtotoolsBuildEnvironment`:
- cache the autotools object with a `_configure_autotools` method
- pass some arguments to the `configure` script (`--enable-shared=yes/no`, `--enable-static=yes/no`)
- move the `install` to `package`

File: recipes/libelfin/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            self.copy("*.dll", dst="bin", src=lib_dir, keep_path=False)
+            self.copy("*.dll.a", dst="lib", src=lib_dir, keep_path=False)
+        else:
+            self.copy("*.a", dst="lib", src=lib_dir, keep_path=False)
+        self.copy("*.lib", dst="lib", src=lib_dir, keep_path=False)
+        # Licenses
+        self.copy("LICENSE", src=self._source_subfolder, dst="licenses")
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
```
Comment (madebr) [inline recipes/libelfin/all/conanfile.py:62]: It looks like the makefile also installs a `.pc` files.
Please set the name of the `pkg_config` generator explicitly.

Comment (klimkin) [reply recipes/libelfin/all/conanfile.py:47]: Switched to external CMakeLists.txt

File: recipes/libelfin/all/conandata.yml (hunk 1 above)
Comment (klimkin) [reply recipes/libelfin/all/conandata.yml]: Removed

File: recipes/libelfin/all/conanfile.py (hunk 2 above)
Comment (klimkin) [reply recipes/libelfin/all/conanfile.py]: Done

**Comment by klimkin** [comment]
@madebr This library requires C++11. What is the strategy to set default cppstd setting?

File: recipes/libelfin/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, "11")
```
Comment (madebr) [inline recipes/libelfin/all/conanfile.py:38]:
```suggestion
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, "11")
```

File: recipes/libelfin/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        cmake = self._configure_cmake()
+        if self.should_build:
+            cmake.build()
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
```
Comment (madebr) [inline recipes/libelfin/all/conanfile.py:57]:
```suggestion
        self.cpp_info.libs = ["elf++", "dwarf++"]
```

File: recipes/libelfin/all/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,67 @@
+cmake_minimum_required(VERSION 3.1)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+project(${CONAN_PACKAGE_NAME} VERSION ${CONAN_PACKAGE_VERSION})
+conan_basic_setup()
```
Comment (madebr) [inline recipes/libelfin/all/CMakeLists.txt:5]:
```suggestion
cmake_minimum_required(VERSION 3.1)
project(cmake_wrapper)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup()
```

Just to be sure that the include can depend on the detected compiler.
e.g. on Apple, the inclusion calls `conan_find_apple_framework`, which calls `find_library`, which is dependent on the current compiler.

File: recipes/libelfin/all/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    OUTPUT source_subfolder/dwarf/to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo '// Automatically generated' > to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo '// DO NOT EDIT' >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo '\#include \"internal.hh\"' >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo 'DWARFPP_BEGIN_NAMESPACE' >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo >> to_string.cc
+    COMMAND python3 ../elf/enum-print.py < dwarf++.hh >> to_string.cc
+    COMMAND python3 ../elf/enum-print.py -s _ -u --hex -x hi_user -x lo_user < data.hh >> to_string.cc
```
Comment (madebr) [inline recipes/libelfin/all/CMakeLists.txt:43]: Don't use `python3`, but [`FindPythonInterp`](https://cmake.org/cmake/help/latest/module/FindPythonInterp.html).
Does the script work on python2? If not, set the correct options and fail if necessary.

File: recipes/libelfin/all/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    VERSION ${CMAKE_PROJECT_VERSION})
+
+install(TARGETS elf++
+    ARCHIVE DESTINATION lib
+    LIBRARY DESTINATION lib
+    PUBLIC_HEADER DESTINATION include/libelfin/elf)
+install(TARGETS dwarf++
+    ARCHIVE DESTINATION lib
+    LIBRARY DESTINATION lib
+    PUBLIC_HEADER DESTINATION include/libelfin/dwarf)
```
Comment (madebr) [inline recipes/libelfin/all/CMakeLists.txt:65]: These miss a `RUNTIME DESTINATION bin` to put the dll for Windows.
Does this package support MSVC dll's?

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch [hunk 10]
```diff
@@ -0,0 +1,217 @@
+diff --git a/dwarf/Makefile b/dwarf/Makefile
+index 71456bd..2d47003 100644
+--- a/dwarf/Makefile
++++ b/dwarf/Makefile
```
Comment (madebr) [inline recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch:4]: I guess you can remove the `Makefile` hunks from the patch, as you use your own `CMakeLists.txt`.

File: recipes/libelfin/all/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    source_subfolder/elf/elf++.hh)
+file(GLOB_RECURSE dwarf_sources source_subfolder/dwarf/*.cc)
+set(dwarf_headers
+    source_subfolder/dwarf/data.hh
+    source_subfolder/dwarf/dwarf++.hh
+    source_subfolder/dwarf/small_vector.hh)
+
+add_custom_command(
+    OUTPUT source_subfolder/elf/to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo '// Automatically generated' > to_string.cc
```
Comment (madebr) [inline recipes/libelfin/all/CMakeLists.txt:20]: I think you are mixing hard tabs with spaces.

File: recipes/libelfin/all/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	COMMAND ${CMAKE_COMMAND} -E echo '\#include \"internal.hh\"' >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo 'DWARFPP_BEGIN_NAMESPACE' >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo >> to_string.cc
+    COMMAND python3 ../elf/enum-print.py < dwarf++.hh >> to_string.cc
+    COMMAND python3 ../elf/enum-print.py -s _ -u --hex -x hi_user -x lo_user < data.hh >> to_string.cc
+	COMMAND ${CMAKE_COMMAND} -E echo 'DWARFPP_END_NAMESPACE' >> to_string.cc
+    DEPENDS source_subfolder/elf/enum-print.py source_subfolder/dwarf/data.hh
+    WORKING_DIRECTORY source_subfolder/dwarf)
+
```
Comment (madebr) [inline recipes/libelfin/all/CMakeLists.txt:47]: re your cppstd question, I would do:
```python
set(CMAKE_CXX_STANDARD 11)
```

Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt:47]: Yes, but I want to enable an override. This worked:
~~~cmake
if(NOT CMAKE_CXX_STANDARD)
    set(CMAKE_CXX_STANDARD 11)
endif()
~~~

Comment (madebr) [reply recipes/libelfin/all/CMakeLists.txt:51]: Looks good.
And if some users sets `compiler.cppstd=98`, then he should not be shooting in his own foot :smile:

Comment (madebr) [reply recipes/libelfin/all/CMakeLists.txt:51]: Never mind, you check for the c++ standard.

File: recipes/libelfin/all/conanfile.py (hunk 6 above)
Comment (klimkin) [reply recipes/libelfin/all/conanfile.py:57]: Why not to use `collect_libs`? With collect_libs we don't have to duplicate the list of installed libraries, it all comes from CMakeLists.txt.

File: recipes/libelfin/all/CMakeLists.txt (hunk 7 above)
Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt]: Thank you for clarification. I would expect the `conanbuildinfo.cmake` to only set variables and have the `conan_*setup()` to run any heavy-weight routines. This way we could feed project name and version to `project()` and make better use of CMAKE_PROJECT_VERSION and CMAKE_PROJECT_NAME variables.

File: recipes/libelfin/all/CMakeLists.txt (hunk 9 above)
Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt]: Fixed.
This is a pure library package, I don't see why it wouldn't support dll's.

File: recipes/libelfin/all/CMakeLists.txt (hunk 11 above)
Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt]: Fixed, thanks.

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch (hunk 10 above)
Comment (klimkin) [reply recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch]: Done

File: recipes/libelfin/all/CMakeLists.txt (hunk 8 above)
Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt]: Yes, Python 3 is required. Fixed.

File: recipes/libelfin/all/CMakeLists.txt (hunk 9 above)
Comment (madebr) [reply recipes/libelfin/all/CMakeLists.txt]: shared libraries on windows uses dll+import libraries. And to export some symbol from a dll (and thus put it in the import library), you need to tag those symbols using `__declspec(dllexport)` when exporting and `__declspec(dllimport)` when using the dll. When using the static library, then NO tags should be supplied.

File: recipes/libelfin/all/conanfile.py (hunk 6 above)
Comment (madebr) [reply recipes/libelfin/all/conanfile.py:57]: I think in this case it's ok, but when libraries are dependent on this `tools.collect_libs` might put them in the wrong order.
And order matters when linking to static libraries.

**Comment by madebr** [comment]
I've asked upstream whether they are open to accepting a cmake build script.
See https://github.com/aclements/libelfin/pull/37#issuecomment-715606804
By combining forces, a better cmake recipe will be created.

File: recipes/libelfin/all/conanfile.py (hunk 6 above)
Comment (klimkin) [reply recipes/libelfin/all/conanfile.py:61]: This makes perfect sense. We have `dwarf++` depends on `elf++`, so then must go in-order. Else basic `cmake` generator` can fail if ordering change. This shouldn't affect if we use `CONAN_PKG::dwarf++`, which still requires explicit dependency. Need better component management to make this right.

**Review by madebr — DISMISSED** [review]

**Comment by madebr** [comment]
Looks like the upstream project owners are silent about cmake support.
So this pr gets a **go** from me.

@SpaceIm @prince-chrismc @uilianries

File: recipes/libelfin/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def build(self):
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            tools.patch(**patch)
+        cmake = self._configure_cmake()
+        if self.should_build:
+            cmake.build()
```
Comment (prince-chrismc) [inline recipes/libelfin/all/conanfile.py:54]: I have never seen this should build before, where does it come from, `ctrl+f` says one result 😕

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+index c780b8e..f77dc0c 100644
+--- a/dwarf/abbrev.cc
++++ b/dwarf/abbrev.cc
+@@ -133,6 +133,10 @@ resolve_type(DW_AT name, DW_FORM form)
+                 case DW_AT::ranges:
+                         return value::type::rangelist;
+ 
++                case DW_AT::lo_user...DW_AT::hi_user:
++                        //HACK: ignore vendor extensions
++                        return value::type::invalid;
```
Comment (prince-chrismc) [inline recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch:11]: Where does this patch come from?

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+diff --git a/elf/elf++.hh b/elf/elf++.hh
+index 562d1de..ee59ed0 100644
+--- a/elf/elf++.hh
++++ b/elf/elf++.hh
+@@ -48,7 +48,7 @@ public:
+  * section data.  Hence, callers must ensure that the loader passed to
+  * this file remains live as long as any such pointer is in use.
+  * Keeping any object that can return such a pointer live is
+- * sufficieint to keep the loader live.
++ * sufficient to keep the loader live.
```
Comment (prince-chrismc) [inline recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch:51]: There's a few changes to the comments in the patch, I am not sure that's the place of C3i

**Review by prince-chrismc — DISMISSED** [review]
It's a soft approval, since I did not take time to deep dive into that top level CMakeLists.txt

File: recipes/libelfin/all/conanfile.py (hunk 13 above)
Comment (SpaceIm) [reply recipes/libelfin/all/conanfile.py:54]: https://docs.conan.io/en/latest/reference/conanfile/attributes.html#should-configure-should-build-should-install-should-test

Can be removed in CCI.

File: recipes/libelfin/all/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+class LibelfinConan(ConanFile):
+    name = "libelfin"
+    description = "C++11 library for reading ELF binaries and DWARFv4 debug information"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/aclements/libelfin"
+    license = "MIT"
+    topics = ("conan", "elf", "dwarf", "libelfin")
+
+    generators = "cmake"
```
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:14]:
```suggestion
```

File: recipes/libelfin/all/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    exports_sources = "CMakeLists.txt", "patches/*"
+
```
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:20]:
```suggestion
    exports_sources = "CMakeLists.txt", "patches/*"
    generators = "cmake"

```

File: recipes/libelfin/all/conanfile.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    _source_subfolder = "source_subfolder"
+    _cmake = None
+
```
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:23]:
```suggestion
    _cmake = None
    
    @property
    def _source_subfolder(self):
        return "source_subfolder"

```

File: recipes/libelfin/all/conanfile.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.configure()
+        return self._cmake
+
```
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:30]:
```suggestion
```

File: recipes/libelfin/all/conanfile.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+        if self.options.shared:
+            del self.options.fPIC
+        if self.settings.compiler.cppstd:
+            tools.check_min_cppstd(self, "11")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
```
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:48]:
```suggestion

    def _configure_cmake(self):
        if self._cmake:
            return self._cmake
        self._cmake = CMake(self)
        self._cmake.configure()
        return self._cmake

```

**Review by SpaceIm — CHANGES_REQUESTED** [review]
Just to make related methods/attributes close to each others:
- cmake generator only make sense because we are exporting a `CMakeLists` wrapper
- `_configure_cmake()` should not be too far from `build()` and `package()`

File: recipes/libelfin/all/conanfile.py (hunk 13 above)
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:54]:
```suggestion
        cmake.build()
```

File: recipes/libelfin/all/conanfile.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        cmake = self._configure_cmake()
+        if self.should_build:
+            cmake.build()
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.libs = ["dwarf++", "elf++"]
```
Comment (SpaceIm) [inline recipes/libelfin/all/conanfile.py:61]:
```suggestion
        self.cpp_info.components["libelf++"].names["pkg_config"] = "libelf++"
        self.cpp_info.components["libelf++"].libs = ["elf++"]
        self.cpp_info.components["libdwarf++"].names["pkg_config"] = "libdwarf++"
        self.cpp_info.components["libdwarf++"].libs = ["dwarf++"]
        self.cpp_info.components["libdwarf++"].requires = ["libelf++"]
```

https://github.com/aclements/libelfin/blob/946dde57e5caef8297b9339f3a7971401d540840/elf/Makefile#L44
https://github.com/aclements/libelfin/blob/946dde57e5caef8297b9339f3a7971401d540840/dwarf/Makefile#L45

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch (hunk 14 above)
Comment (klimkin) [reply recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch:11]: I think there is a better way to do this...
coz recommends using libelfin v0.2, but the only libelfin version I found is working with coz is the latest. Unfortunately, it wasn't properly released. There are a few commits on top of libelfin v0.3. So the patch does the fix for some files to that commit. I can change the recipe to use git.clone to get the latest, but what would I use for the version? Using the `master` from libelfin will have an issue with build reproducibility, so we can use specific hash. But how do you choose the version tag for it? `0.3-1`?

File: recipes/libelfin/all/conanfile.py (hunk 18 above)
Comment (klimkin) [reply recipes/libelfin/all/conanfile.py:23]: I would use `@property` if there was any _behavior_. For example, if value depends on options or settings. For a plain value the original code is all it needs.

File: recipes/libelfin/all/conanfile.py (hunk 13 above)
Comment (klimkin) [reply recipes/libelfin/all/conanfile.py:54]: `cmake.build()` handles `self.should_build` internally. I will remove this.

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch (hunk 15 above)
Comment (klimkin) [reply recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch:51]: Related to the question about the patch above.

File: recipes/libelfin/all/conanfile.py (hunk 18 above)
Comment (prince-chrismc) [reply recipes/libelfin/all/conanfile.py:23]: I understand your point of view and you are right.
However, for the sake of consistency, usually the cached cmake is first... it's harmless to apply the suggestion if you dont mind 🙏

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch (hunk 14 above)
Comment (prince-chrismc) [reply recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch:11]: Perhaps just take those commit(s) and apply them and reference them as a comment on top of the patch?

we have "virtual versioning" which is `cci.yyyymmdd` but I dont think we need that here, try the suggestion above?

File: recipes/libelfin/all/conanfile.py (hunk 18 above)
Comment (klimkin) [reply recipes/libelfin/all/conanfile.py]: Done!

File: recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch (hunk 14 above)
Comment (klimkin) [reply recipes/libelfin/all/patches/946dde57e5caef8297b9339f3a7971401d540840.patch]: Reduced patch to the critical fixes and added comment:
~~~
> head patches/9d0db16d0a0b3c4f8aaa60a3e4dab295df34b6b2.patch
#
# Updates v0.3 to the commit 9d0db16d0a0b3c4f8aaa60a3e4dab295df34b6b2:
#   9d0db16 fix data initialization of elf segment
#   74f1116 ignore dwarf extensions of gcc in the line table
#   ac45a09 Added #include <errno.h> to files that use errno.
#   4a678c7 Don't mandate that compilation units need a DW_AT_comp_dir for line table parsing
#   85812f6 Fix typos
#
~~~

**Comment by klimkin** [comment]
@uilianries, @danimtb Could you review, please? So I can continue with `coz` recipe.

**Review by danimtb — APPROVED** [review]

**Comment by SpaceIm** [comment]
@klimkin What did you change in this last rebase? Now it's very hard to know if all review comments was taken into account.
PR are merge squashed anyway.

**Comment by klimkin** [comment]
> @klimkin What did you change in this last rebase? Now it's very hard to know if all review comments was taken into account.
> PRs are squash merged anyway.

@SpaceIm I apologize, next time I will keep review-related commits.

The only last changes were replacing patch file with a cleaner version and adding comments to the patch file, as requested by  @prince-chrismc.

File: recipes/libelfin/all/conanfile.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    _source_subfolder = "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.settings.compiler == "Visual Studio":
+            raise ConanInvalidConfiguration("libelfin doesn't support compiler: {} on OS: {}.".
+                                            format(self.settings.compiler, self.settings.os))
```
Comment (madebr) [inline recipes/libelfin/all/conanfile.py:31]:
```suggestion
        if self.settings.os == "Windows":
            raise ConanInvalidConfiguration("libelfin doesn't support Windows")
```

I just tested on mingw@Linux, and it failed with:
```
/home/maarten/.conan/data/libelfin/0.3/_/_/build/5f8894dbe4d6e2b6274767c0bd1003242682f09a/source_subfolder/elf/mmap_loader.cc:11:10: fatal error: sys/mman.h: No such file or directory
 #include <sys/mman.h>
          ^~~~~~~~~~~~
compilation terminated.
```

**Review by prince-chrismc — APPROVED** [review]

**Comment by klimkin** [comment]
@SSE4 , @uilianries Could you approve, please?

**Comment by prince-chrismc** [comment]
@klimkin Don't forget the community reviewers we count as well! #2857

**Review by SSE4 — DISMISSED** [review]

**Comment by prince-chrismc** [comment]
This PR is ready for review 😃 

@madebr
@SpaceIm
@ericLemanissier
@Croydon
@intelligide
@theirix
@gocarlos

**Review by madebr — DISMISSED** [review]

File: recipes/libelfin/all/conandata.yml [hunk 23]
```diff
@@ -0,0 +1,8 @@
+sources:
+  "0.3":
+    url: "https://github.com/aclements/libelfin/archive/v0.3.tar.gz"
+    sha256: "c338942b967582922b3514b54b93175ca9051a9668db92dd8ef619824d443ac7"
+patches:
+  "0.3":
+    - patch_file: "patches/9d0db16d0a0b3c4f8aaa60a3e4dab295df34b6b2.patch"
```
Comment (intelligide) [inline recipes/libelfin/all/conandata.yml:7]: Can you provide an explicit file name ?

**Review by intelligide — CHANGES_REQUESTED** [review]

File: recipes/libelfin/all/conandata.yml (hunk 23 above)
Comment (klimkin) [reply recipes/libelfin/all/conandata.yml:7]: The patch directly related to commit `9d0db16d0a0b3c4f8aaa60a3e4dab295df34b6b2` from the original repository. The commit goes on top of the release and there was no release available.
Will something like `commit-9d0db16d0a0b3c4f8aaa60a3e4dab295df34b6b2.patch` work better?

File: recipes/libelfin/all/test_package/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+if(NOT CMAKE_CXX_STANDARD)
+    set(CMAKE_CXX_STANDARD 11)
+endif()
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (intelligide) [inline recipes/libelfin/all/test_package/CMakeLists.txt:10]:
```suggestion
```

File: recipes/libelfin/all/test_package/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (intelligide) [inline recipes/libelfin/all/test_package/CMakeLists.txt:13]:
```suggestion
target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
set_target_properties(${PROJECT_NAME} PROPERTIES
    CXX_STANDARD 11
    CXX_STANDARD_REQUIRED ON)
```

**Review by intelligide — CHANGES_REQUESTED** [review]

File: recipes/libelfin/all/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+install(TARGETS elf++
+    ARCHIVE DESTINATION lib
+    LIBRARY DESTINATION lib
+    RUNTIME DESTINATION bin
+    PUBLIC_HEADER DESTINATION include/libelfin/elf)
+install(TARGETS dwarf++
+    ARCHIVE DESTINATION lib
+    LIBRARY DESTINATION lib
+    RUNTIME DESTINATION bin
+    PUBLIC_HEADER DESTINATION include/libelfin/dwarf)
```
Comment (intelligide) [inline recipes/libelfin/all/CMakeLists.txt:74]:
```suggestion
install(TARGETS elf++
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
    PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/libelfin/elf)
install(TARGETS dwarf++
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
    PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/libelfin/dwarf)
```

File: recipes/libelfin/all/conandata.yml (hunk 23 above)
Comment (intelligide) [reply recipes/libelfin/all/conandata.yml:7]: Yes, prefixing it with `commit-` is much better.

File: recipes/libelfin/all/test_package/CMakeLists.txt (hunk 25 above)
Comment (klimkin) [reply recipes/libelfin/all/test_package/CMakeLists.txt:13]: This seems to hard-code c++11 as the standard, but we only want to set a minimal version. So in original code Conan can override and compile with c++17 and later.

**Review by intelligide — CHANGES_REQUESTED** [review]

File: recipes/libelfin/all/CMakeLists.txt (hunk 26 above)
Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt:74]: `CMAKE_INSTALL_*` variables are available only with GNUInstallDirs module. Why to use it? It's wordy and doing just the same.

Looking at other recipes I don't see there is a standard. Need some justification for the suggestion.

Comment (madebr) [reply recipes/libelfin/all/CMakeLists.txt:74]: For a recipe it's ok to hardcode the paths I guess.
Have you looked in upstreaming these patches?
Then @intelligide 's suggestion + `include(GNUInstallDirs)` would be super.

File: recipes/libelfin/all/conandata.yml (hunk 23 above)
Comment (klimkin) [reply recipes/libelfin/all/conandata.yml]: https://github.com/conan-io/conan-center-index/pull/3283/commits/1bce461afd70453f4a3ac2ece2c44bc8d7eef104

File: recipes/libelfin/all/CMakeLists.txt (hunk 26 above)
Comment (intelligide) [reply recipes/libelfin/all/CMakeLists.txt:74]: Conan already set these variables to the default recipe paths. (See [here](https://github.com/conan-io/conan/blob/edb3a4a1f7739027c0665749b82b1cc473453ab6/conans/client/build/cmake_flags.py#L356)). Adding `include(GNUInstallDirs)` is better.

Comment (klimkin) [reply recipes/libelfin/all/CMakeLists.txt:74]: The original repository doesn't have CMake files. Yes, we contacted the authors of the library about adding CMake support, but we didn't get any reply so far.

File: recipes/libelfin/all/test_package/CMakeLists.txt (hunk 25 above)
Comment (klimkin) [reply recipes/libelfin/all/test_package/CMakeLists.txt:13]: But I like, it would be attached as a property.

**Review by intelligide — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Please refrain from force pushing, it forces us to review the entire PR ... which was not my intention!

**Comment by klimkin** [comment]
> Please refrain from force pushing, it forces us to review the entire PR ... which was not my intention!

Sorry, I realized this too late, need to unlearn this habit.

**Review by SSE4 — APPROVED** [review]

**Pull request merged**